apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource              `json:"source,omitempty"`
+	Reconcile []StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
+}
+
+// StackDefinitionBehaviorCRD represents the CRD which the stack's behavior controller will watch. When CRs of this
+// CRD kind appear and are modified in the cluster, the behavior will execute.
+type StackDefinitionBehaviorCRD struct {
+	APIVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+}
+
+// StackResourceEngineConfiguration represents a configuration for a resource engine, such as helm2 or kustomize.
+type StackResourceEngineConfiguration struct {
```

Comment:
This will probably introduce `sigs.k8s.io/kustomize/api v0.3.0` dependency as we need to let people define their `Kustomization` object in the template stack CRD. Well, that's actually the case no matter what since we'll have to embed `Kustomization` anyway.

In the future, if we go with [separating `TemplateStack` into its own stack repo](https://github.com/crossplaneio/crossplane/issues/1209#issuecomment-579663571), this should be alleviated but for now it seems fine.

Reply:
> We can possibly have something similar. What do you think?

I agree with this. This is about what I was imagining anyway, but the pattern you linked is cleaner and I wasn't aware of it. Thanks for calling this out!

> This will probably introduce sigs.k8s.io/kustomize/api v0.3.0 dependency

Yeah, that's true. I am also okay with that happening for now and us reassessing in the future.

Reply:
I stand corrected; as you pointed in Slack `Kustomization` object cannot be used as a CRD field type since it doesn't have DeepCopy methods. Using `*unstructured.Unstructured` and later [making the conversion in Go code](https://github.com/crossplaneio/resourcepacks/pull/2/files#diff-7ddfb3e035b42cd70649cc33393fe32cR89) makes sense to me.

Reply:
Yeah, I think that's the easiest for now. I was originally thinking we could define our own DeepCopy at some point, but after some research I learned that isn't entirely possible since [type methods must be declared in the same package as the type](https://tour.golang.org/methods/3). I feel it'd probably be worth investigating whether kustomize plans to add that, and asking them about it